Adopt the dockerfile-language-server-nodejs npm module

rcjsuen opened the pull request:
This PR resolves issue #99 by creating the necessary changes to `package.json` and `dockerExtension.ts` to get the dockerfile-language-server-nodejs npm module integrated into the vscode-docker extension.

Here is how you can test this PR:

1. Pull this PR into your repository.
2. Invoke `npm install` to install the Dockerfile LSP server. This should be relatively quick.
3. Debug > Launch Extension
4. Cross your fingers, open a Dockerfile and see if the installation has completed successfully!

Here is a sample Dockerfile that you can use to get a sense of what is currently possible with this language server. Follow the comments and have fun! Note that you should probably turn off `docker.enableLinting` to more clearly see what the server does and does not provide in terms of validation.

```Dockerfile
#escape=a
# click in the above directive error to convert 'a' to a backslash or backtick
# use Ctrl+Shift+O 'Go to Symbol in File...' to get an overview of the document
FROM alpine AS setup
# invoke completion here to check that the
# dockerhub image suggestions are still displayed
FROM 
ENV var=value
# error here because ENV must have two arguments
ENV var2

# should get a warning about EXPOSE not being written in uppercase
# click in the warning to get a code action to convert it to uppercase
expose 8081

# invoke completion here to get possible ONBUILD trigger instructions
ONBUILD 

# hover over $var to get 'value'
# click in $var and use 'Go to Definition' to jump to the ENV statement
RUN echo $var

# click in setup and use 'Go to Definition' to jump to the FROM statement
COPY --from=setup out/ out/

# click in $var2 to get document highlighting of the references
# try renaming var2 to var3
RUN echo $var2 \
# try formatting the document or a selection here to test newline escape detection
		echo $var2 \
				echo $var2 \
echo $var2 \
				echo $var2 \
		echo $var2
```

msftclas commented:
__@rcjsuen__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

chrisdias commented:
Hi!  Thanks for making this PR, I'm looking forward to merging it in :). I've been testing this out today and in a simple Dockerfile I'm getting a lot of "Invalid Argument" warnings. Typically I get it after inserting a COPY or ARG statement, and then I get them pretty much on every line.

I have not had a chance to narrow down where they are coming from, but I don't see them when running the extension without this change. Can you take a look?

rcjsuen commented:
Hi @chrisdias, thanks for taking a look.

Do you have a screenshot of the error message and the state of the editor?
I don't have any error messages that say "invalid argument". I do have ones that say "invalid containerPort" for `EXPOSE` and "invalid signal" for `STOPSIGNAL`. Seeing the editor and your Dockerfile should hopefully help me narrow it down.

rcjsuen commented:
[Here](https://github.com/rcjsuen/dockerfile-language-server-nodejs/blob/1e3fc462bc6cfeeea5079822567a3a1dcbcfd192/src/dockerValidator.ts#L251-L271) is a list of all the validation messages that exist in the 0.0.1 release of the server.

chrisdias commented:
@rcjsuen I tested this a bit on Friday after removing other extensions I had installed and I did not see the "Invalid Argument" warnings anymore. I'll do some more testing this week and hopefully we can merge this in!  

@lostintangent can you test as well?

lostintangent commented:
@chrisdias Yep! I'm very excited to get this merged as well, so I'll test it thoroughly later this afternoon.

rcjsuen commented:
Thanks you two! Please do not hesitate to reach out again. If you see anything fishy and the classic software workarounds of "opening and closing the editor" or "restarting the extension" doesn't seem to resolve it, please _upload_ the Dockerfile and I'll take a look. Newline characters may be playing a role so I'd prefer an upload over a copy/pasted comment on GitHub. Thanks! :)

lostintangent commented:
@rcjsuen This looks really great at first use! I wasn't able to get the DockerHub image name suggestions to work, but I'll do a little more testing tomorrow morning, since it may have been my machine. Everything else was working beautifully.

rcjsuen commented:
Thanks, @lostintangent.

I deleted my `node_modules/dockerfile-language-server-nodejs` folder just now and invoked `npm install` from scratch and then fired up the debugger. I reproduced the problem but when I switched tabs back to my original VS Code I realized that the debugger had paused somewhere in `bootstrap.js`. I restarted the extension and tried again and it worked. Perhaps there is some startup sequencing thing that I'm not considering in my code?

rcjsuen commented:
Some conflicts were introduced from 3a33301d6bd6b638960392d67cd8c16bdbbb5b00 in #105 so I resolved them and force pushed my branch for a clean slate.

lostintangent commented:
@rcjsuen I unfortunately still haven't been able to get the DockerHub image completion working, so I need to try your workaround steps again to see if I can get it working.

chrisdias commented:
@lostintangent can we take this PR and (temporarily) lose completion against dockerhub (assuming it doesn't throw)?  it doesn't work reliably enough today anyway, and I really want to move to this language server...  then we can make another PR/fix to get completion against registries to work properly.

lostintangent commented:
@chrisdias Yep! That plan sounds great to me.

rcjsuen commented:
Hi you two, thanks again for taking the time to look at this.

I wonder if there is a way to tell VS Code to use an LSP except for its code completion feature? That would allow you to get the same old completion behaviour as before via the handwritten code but leverage the language server for other stuff like formatting and variable highlighting. I don't think so based on what I see in [vscode-languagesever-node](https://github.com/Microsoft/vscode-languageserver-node/blob/master/client/src/client.ts#L1411-L1418) since the `ClientCapabilities` is completely hard coded but I could be wrong.

For my personal roadmap, providing support for LSP features by leveraging the Docker Registry HTTP API is certainly something I am interested in implementing in my language server. Unfortunately, search support for image names seems to be nowhere in sight in the v2 API as you can see from [docker/distribution#206](https://github.com/docker/distribution/issues/206) and its many, many duplicates.

I'm a little busy right now so it may take me some time to resolve the conflicts for this PR but I will try to remain responsive with regards to any other comments you may have about this PR.

Thank you for your understanding! :)

chrisdias commented:
@rcjsuen are you ok with us fixing the merge conflicts and bringing this in?  We can address the registry query after...

rcjsuen commented:
@chrisdias Not a problem. If either of you have the time to resolve the conflict, that would be great and I have no qualms with that. Thanks! :)

chrisdias commented:
merged with master
@lostintangent if you are ok with this i'll bring this up...

rcjsuen commented:
Hey guys, I've resolved the conflict and force pushed it to my `docker-lsp` branch again.

lostintangent commented:
@rcjsuen Perfecr, thanks so much!
